Stripe/STPAddressViewModel.m

```diff
@@ -116,72 +116,82 @@ - (instancetype)initWithSEPADebitFields {
+    [self updatePostalCodeFieldIfNecessary];
 }
 
-- (void)updatePostalCodeCellIfNecessary {
-    if (self.addressType == STPAddressTypeSEPADebit) {
+- (void)updatePostalCodeFieldIfNecessary {
+    if (![self displaysPostalCodeField]) {
         return;
     }
     STPPostalCodeType postalCodeType = [STPPostalCodeValidator postalCodeTypeForCountryCode:_addressFieldTableViewCountryCode];
     BOOL shouldBeShowingPostalCode = (postalCodeType != STPCountryPostalCodeTypeNotRequired);
+    // Add postal code field
     if (shouldBeShowingPostalCode && !self.showingPostalCodeCell) {
-        if (self.addressType == STPAddressTypeBilling && self.requiredBillingAddressFields == STPBillingAddressFieldsZip) {
-            self.addressCells = @[
-                                  [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeZip contents:@"" lastInList:YES delegate:self]
-                                  ];
-            [self.delegate addressViewModel:self addedCellAtIndex:0];
-            [self.delegate addressViewModelDidChange:self];
-        }
-        else if (self.containsStateAndPostalFields) {
-            // Add after STPAddressFieldTypeState
-            NSUInteger stateFieldIndex = [self.addressCells indexOfObjectPassingTest:^BOOL(STPAddressFieldTableViewCell * _Nonnull obj, NSUInteger __unused idx, BOOL * _Nonnull __unused stop) {
-                return (obj.type == STPAddressFieldTypeState);
+        NSNumber *previousFieldType = [self fieldTypeAbovePostalCode];
+        NSUInteger previousFieldIndex = NSNotFound;
+        if (previousFieldType) {
+            previousFieldIndex = [self.addressCells indexOfObjectPassingTest:^BOOL(STPAddressFieldTableViewCell *obj, NSUInteger __unused idx, BOOL *__unused stop) {
+                return (obj.type == [previousFieldType integerValue]);
             }];
+        }
 
-            if (stateFieldIndex != NSNotFound) {
-                NSUInteger zipFieldIndex = stateFieldIndex + 1;
-
-                NSMutableArray<STPAddressFieldTableViewCell *> *mutableAddressCells = self.addressCells.mutableCopy;
-                [mutableAddressCells insertObject:[[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeZip contents:@"" lastInList:NO delegate:self]
-                                          atIndex:zipFieldIndex];
-                self.addressCells = mutableAddressCells.copy;
-                [self.delegate addressViewModel:self addedCellAtIndex:zipFieldIndex];
-                [self.delegate addressViewModelDidChange:self];
-            }
+        STPAddressFieldTableViewCell *zipCell = [[STPAddressFieldTableViewCell alloc] initWithType:STPAddressFieldTypeZip contents:@"" lastInList:NO delegate:self];
+        NSUInteger zipFieldIndex;
+        if (previousFieldIndex != NSNotFound) {
+            zipFieldIndex = previousFieldIndex + 1;
+        } else {
+            zipFieldIndex = 0;
+            zipCell.lastInList = YES;
```

Comment:
I dont think this logic is right? Just because its the first thing doesnt mean its also the last, and vice versa.

Also if being added as last in list and there are already other fields, need to alter the previous last in list's value to NO